### PR TITLE
feat: support multiple themes

### PR DESCRIPTION
**Pull request by Sebastian-Haehnlein**
## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] CI-related changes
- [ ] Documentation content changes
- [ ] application / infrastructure changes
- [ ] Other... Please describe:

## What Is the Current Behavior?

Issue Number: 
https://github.com/intershop/intershop-pwa/issues/49

## What Is the New Behavior?
multiple styles & matrix-parameter switch

## Does this PR Introduce a Breaking Change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: angular.json [hunk 1]
```diff
@@ -31,10 +31,25 @@
...
+                "input": "src/theme/themes/blue/style.scss",
+                "lazy": true,
+                "bundleName": "blue"
+              }
+            ],
             "scripts": []
           },
           "configurations": {
+            "mock": {
+              "extractCss": true
```
Comment (dhhyi) [inline angular.json]: why not add this directly in the builder options above? That way no additional configuration would be required

File: src/environments/environment.model.ts [hunk 2]
```diff
@@ -70,4 +72,7 @@ export interface Environment {
 
   // configuration of the available locales - hard coded for now
   locales: { lang: string; currency: string; value: string; displayName: string; displayLong: string }[];
+
+  // themes
+  themes: Theme[];
```
Comment (dhhyi) [inline src/environments/environment.model.ts]: I'd rather not see the `Theme` model being used here. We're not doing it with `Locale` either...

File: src/app/core/store/configuration/configuration.effects.ts [hunk 3]
```diff
@@ -49,11 +50,14 @@ export class ConfigurationEffects {
       this.stateProperties.getStateOrEnvOrDefault<string>('ICM_APPLICATION', 'icmApplication'),
       this.stateProperties
         .getStateOrEnvOrDefault<string | string[]>('FEATURES', 'features')
-        .pipe(map(x => (typeof x === 'string' ? x.split(/,/g) : x)))
+        .pipe(map(x => (typeof x === 'string' ? x.split(/,/g) : x))),
+      this.stateProperties
+        .getStateOrEnvOrDefault<string>('THEME', 'themes')
+        .pipe(map(x => (typeof x === 'string' ? x : (x as Theme[]).find(t => t.default).name)))
```
Comment (dhhyi) [inline src/app/core/store/configuration/configuration.effects.ts]: how about moving the type:
```
        .getStateOrEnvOrDefault<string | Theme[]>('THEME', 'themes')
        .pipe(map(x => (typeof x === 'string' ? x : x.find(t => t.default).name)))
```

File: src/app/core/facades/theme.facade.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  constructor(
+    private rendererFactory: RendererFactory2,
+    @Inject(DOCUMENT) private document: Document,
+    @Inject(THEMES) private themes: Theme[],
+    store: Store<{}>
+  ) {
+    this.head = this.document.head;
+    this.renderer = this.rendererFactory.createRenderer(undefined, undefined);
+    this.theme$ = new BehaviorSubject(this.themes.find(x => x.default).name);
```
Comment (dhhyi) [inline src/app/core/facades/theme.facade.ts]: how about initializing the `theme$` with `new ReplaySubject(1)` and just `next`ing the first value here?

File: src/app/core/facades/theme.facade.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    store.pipe(select(getTheme)).subscribe(theme => {
```
Comment (dhhyi) [inline src/app/core/facades/theme.facade.ts]: I think you can directly use the Observable above here: `.subscribe(this.thene$)`. That way you shouldn't neet the `set` anymore.

File: src/app/core/facades/theme.facade.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      this.set(theme);
+    });
+
+    this.theme$.pipe(distinctUntilChanged()).subscribe(async theme => {
+      await this.loadCss(`${theme}.css`);
+      if (this.themeLinks.length === 2) {
```
Comment (dhhyi) [inline src/app/core/facades/theme.facade.ts]: Please add a comment. This removes a theme when it is already set, right?

File: src/app/app.component.ts [hunk 7]
```diff
@@ -24,7 +25,11 @@ export class AppComponent implements OnInit {
   wrapperClasses$: Observable<string[]>;
   deviceType$: Observable<DeviceType>;
 
-  constructor(private appFacade: AppFacade, @Inject(PLATFORM_ID) platformId: string) {
+  constructor(
+    private appFacade: AppFacade,
+    @Inject(PLATFORM_ID) platformId: string,
+    public themeFacade: ThemeFacade // just for init (constructor)
```
Comment (dhhyi) [inline src/app/app.component.ts]: How about providing it in `ConfigurationModule` instead.

File: src/app/core/facades/theme.facade.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import { distinctUntilChanged } from 'rxjs/operators';
+
+import { THEMES } from 'ish-core/configurations/injection-keys';
+import { Theme } from 'ish-core/models/theme/theme.model';
+import { getTheme } from 'ish-core/store/configuration';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ThemeFacade {
```
Comment (dhhyi) [inline src/app/core/facades/theme.facade.ts]: this is not really a facade, it is more something like a utility service. please rename and move it

File: src/environments/environment.model.ts (hunk 2 above)
Comment (Sebastian-Haehnlein) [reply src/environments/environment.model.ts]: decision for typed environments-model

File: src/app/core/facades/theme.facade.ts (hunk 5 above)
Comment (Sebastian-Haehnlein) [reply src/app/core/facades/theme.facade.ts]: theme$ removed, state used

File: .gitlab-ci.yml [hunk 9]
```diff
@@ -397,6 +397,7 @@ deploy_demo_nginx:
       -e PWA_2_SUBDOMAIN=b2b
       -e PWA_2_CHANNEL=inSPIRED-inTRONICS_Business-Site
       -e PWA_2_FEATURES=quoting,recently,compare,businessCustomerRegistration,advancedVariationHandling,sentry
+      -e PWA_2_THEME=blue
```
Comment (dhhyi) [inline .gitlab-ci.yml]: I wanna see the blue and I#m not seein it?! https://www.youtube.com/watch?v=eJkbKXTjNrA
C'est bleu? Non estas azure! :rofl:

File: src/environments/environment.model.ts [hunk 10]
```diff
@@ -63,11 +67,14 @@ export interface Environment {
...
 
   // enable or disable service worker
   serviceWorker: boolean;
 
   // configuration of the available locales - hard coded for now
-  locales: { lang: string; currency: string; value: string; displayName: string; displayLong: string }[];
+  locales: Locale[];
+
+  // themes
+  themes: Theme[];
```
Comment (dhhyi) [inline src/environments/environment.model.ts]: As discussed: let's build it back to a single string. ;)

File: src/app/core/store/configuration/configuration.effects.ts [hunk 11]
```diff
@@ -49,11 +50,14 @@ export class ConfigurationEffects {
       this.stateProperties.getStateOrEnvOrDefault<string>('ICM_APPLICATION', 'icmApplication'),
       this.stateProperties
         .getStateOrEnvOrDefault<string | string[]>('FEATURES', 'features')
-        .pipe(map(x => (typeof x === 'string' ? x.split(/,/g) : x)))
+        .pipe(map(x => (typeof x === 'string' ? x.split(/,/g) : x))),
+      this.stateProperties
+        .getStateOrEnvOrDefault<string | Theme[]>('THEME', 'themes')
+        .pipe(map(x => (typeof x === 'string' ? x : x.find(t => t.default).name)))
```
Comment (dhhyi) [inline src/app/core/store/configuration/configuration.effects.ts]: please also rename theme `orange` to `default` and fallback to it here, if the incoming property is falsy.

File: src/theme/components/header/header.scss [hunk 12]
```diff
@@ -165,8 +165,20 @@ header {
   }
 }
 
+.logo {
+  display: inline-block;
+  width: 114px;
+  height: 40px;
+  background: url('/assets/img/logo.png') center;
+}
```
Comment (dhhyi) [inline src/theme/components/header/header.scss]: sticky header logo seems odd: 
image

**Comment by dhhyi** [comment]
> We still need to wait for the definition of the new B2B color and the fitting logo images.

Might I suggest doing that later in another PR? There is plenty of business value already packed in here, we can release it for projects to use and also provide schematics creating a new theme here. (see #77)

**Pull request merged**